remote.go

```diff
@@ -17,33 +17,45 @@ import (
 	"gopkg.in/src-d/go-git.v4/storage/memory"
 )
 
-var NoErrAlreadyUpToDate = errors.New("already up-to-date")
+var (
+	NoErrAlreadyUpToDate = errors.New("already up-to-date")
+	ErrAlreadyConnected  = errors.New("remote already connected")
+	ErrFetchNotConnected = errors.New("fetch not connected")
+	ErrPushNotConnected  = errors.New("push not connected")
+)
 
 // Remote represents a connection to a remote repository
 type Remote struct {
 	c *config.RemoteConfig
+	r *Repository
 	s Storer
 	p sideband.Progress
 
 	// cache fields, there during the connection is open
 	endpoint     transport.Endpoint
 	client       transport.Client
 	fetchSession transport.FetchPackSession
+	sendSession  transport.SendPackSession
+	session      transport.Session
```

Comment:
session can be either fetchSession / sendSession. It's used for some common operations.